sale_invoice_plan/models/sale.py

```diff
@@ -0,0 +1,253 @@
+            ("sent", "Quotation Sent"),
+            ("sale", "Sales Order"),
+            ("done", "Locked"),
+            ("cancel", "Cancelled"),
+        ],
+        string="Status",
+        related="sale_id.state",
+        store=True,
+        index=True,
+    )
+    installment = fields.Integer(string="Installment")
+    plan_date = fields.Date(string="Plan Date", required=True)
+    invoice_type = fields.Selection(
+        [("advance", "Advance"), ("installment", "Installment")],
+        string="Type",
+        required=True,
+        default="installment",
+    )
+    last = fields.Boolean(
+        string="Last Installment",
+        compute="_compute_last",
+        help="Last installment will create invoice use remaining amount",
+    )
+    percent = fields.Float(
+        string="Percent",
+        digits="Product Unit of Measure",
+        help="This percent will be used to calculate new quantity",
+    )
+    invoice_move_ids = fields.Many2many(
+        "account.move",
+        relation="sale_invoice_plan_invoice_rel",
+        column1="plan_id",
+        column2="move_id",
+        string="Invoices",
+        readonly=True,
+    )
+    to_invoice = fields.Boolean(
+        string="Next Invoice",
+        compute="_compute_to_invoice",
+        help="If this line is ready to create new invoice",
+    )
+    invoiced = fields.Boolean(
+        string="Invoice Created",
+        compute="_compute_invoiced",
+        help="If this line already invoiced",
+    )
+    _sql_constraint = [
+        (
+            "unique_instalment",
+            "UNIQUE (sale_id, installment)",
+            "Installment must be unique on invoice plan",
+        )
+    ]
+
+    def _compute_to_invoice(self):
+        """If any invoice is in draft/open/paid do not allow to create inv.
+        Only if previous to_invoice is False, it is eligible to_invoice.
+        """
+        for rec in self:
+            rec.to_invoice = False
+        for rec in self.sorted("installment"):
+            if rec.state != "sale":  # Not confirmed, no to_invoice
+                continue
+            if not rec.invoiced:
+                rec.to_invoice = True
+                break
+
+    def _compute_invoiced(self):
+        for rec in self:
+            invoiced = rec.invoice_move_ids.filtered(
+                lambda l: l.state in ("draft", "posted")
+            )
+            rec.invoiced = invoiced and True or False
+
+    def _compute_last(self):
+        for rec in self:
+            last = max(rec.sale_id.invoice_plan_ids.mapped("installment"))
+            rec.last = rec.installment == last
+
+    def _compute_new_invoice_quantity(self, invoice_move):
+        self.ensure_one()
+        if self.last:  # For last install, let the system do the calc.
+            return
+        percent = self.percent
+        move = invoice_move.with_context({"check_move_validity": False})
+        for line in move.invoice_line_ids:
+            assert (
```

Comment:
We are in an Odoo context, better to raise UserError instead

Reply:
Okay, thanks